agents/postgres/pgstatmonitor/stat_monitor_cache.go

```diff
@@ -0,0 +1,257 @@
+	d := s.newest.Sub(s.oldest)
+	return fmt.Sprintf("current=%d: updated=%d added=%d removed=%d; %s - %s (%s)",
+		s.current, s.updatedN, s.addedN, s.removedN,
+		s.oldest.UTC().Format("2006-01-02T15:04:05Z"), s.newest.UTC().Format("2006-01-02T15:04:05Z"), d)
+}
+
+// statMonitorCache provides cached access to pg_stat_monitor.
+// It retains data longer than this table.
+type statMonitorCache struct {
+	retain time.Duration
+	l      *logrus.Entry
+
+	rw       sync.RWMutex
+	items    map[int64]*pgStatMonitorExtended
+	added    map[int64]time.Time
+	updatedN uint
+	addedN   uint
+	removedN uint
+}
+
+// newStatMonitorCache creates new statMonitorCache.
+func newStatMonitorCache(retain time.Duration, l *logrus.Entry) *statMonitorCache {
+	return &statMonitorCache{
+		retain: retain,
+		l:      l,
+		items:  make(map[int64]*pgStatMonitorExtended),
+		added:  make(map[int64]time.Time),
+	}
+}
+
+// getStatMonitorExtended returns the current state of pg_stat_monitor table with extended information (database, username, tables)
+// and the previous cashed state.
+func (ssc *statMonitorCache) getStatMonitorExtended(ctx context.Context, q *reform.Querier) (current, prev map[int64]*pgStatMonitorExtended, err error) {
+	var totalN, newN, newSharedN, oldN int
+	start := time.Now()
+	defer func() {
+		dur := time.Since(start)
+		ssc.l.Debugf("Selected %d rows from pg_stat_monitor in %s: %d new (%d shared tables), %d old.", totalN, dur, newN, newSharedN, oldN)
+	}()
+
+	ssc.rw.RLock()
+	current = make(map[int64]*pgStatMonitorExtended, len(ssc.items))
+	prev = make(map[int64]*pgStatMonitorExtended, len(ssc.items))
+	for k, v := range ssc.items {
+		prev[k] = v
+	}
+	ssc.rw.RUnlock()
+
+	// load all databases and usernames first as we can't use querier while iterating over rows below
+	databases := queryDatabases(q)
+	usernames := queryUsernames(q)
+
+	// the same query can appear several times (with different database and/or username),
+	// so cache results of the current iteration too
+	tables := make(map[int64][]string)
+
+	rows, err := q.SelectRows(pgStatMonitorView, "WHERE queryid IS NOT NULL AND query IS NOT NULL")
+	if err != nil {
+		err = errors.Wrap(err, "failed to query pg_stat_monitor")
+		return
+	}
+	defer rows.Close() //nolint:errcheck
+
+	for ctx.Err() == nil {
+		var row pgStatMonitor
+		if err = q.NextRow(&row, rows); err != nil {
+			if err == reform.ErrNoRows {
+				err = nil
+			}
+			break
+		}
+		totalN++
+
+		c := &pgStatMonitorExtended{
+			pgStatMonitor: row,
+			Database:      databases[row.DBID],
+			Username:      usernames[row.UserID],
+		}
+
+		if p := prev[c.QueryID]; p != nil {
+			oldN++
+
+			// use previous values
+			c.Tables = p.Tables
+			c.Query, c.IsQueryTruncated = p.Query, p.IsQueryTruncated
+		} else {
+			newN++
+
+			// do not extract tables again if we saw this query during this iteration already
+			if tables[c.QueryID] == nil {
+				tables[c.QueryID] = extractTables(c.Query, ssc.l)
```

Comment:
we don't need to use extractTables in pg stat monitor since we already have them in the view.

Reply:
@JiriCtvrtka ^